d2l-table.html

```diff
@@ -151,21 +153,120 @@
 			}
 		</style>
 		<d2l-scroll-wrapper show-actions>
-			<slot></slot>
+			<slot id="slot"></slot>
 		</d2l-scroll-wrapper>
 	</template>
 	<script>
 		Polymer({
 			is: 'd2l-table',
+			behaviors: [
+				D2L.PolymerBehaviors.Table.LocalObserverBehavior
+			],
+			listeners: {
+				'd2l-table-local-observer': '_handleLocalObserver'
+			},
 			properties: {
 				role: {
 					type: String,
 					value: 'table',
 					reflectToAttribute: true
 				}
 			},
+			__applyInQueue: false,
+			attached: function() {
+				Polymer.RenderStatus.afterNextRender(this, function() {
+					this._applyClassNames();
+				}.bind(this));
+			},
 			notifyResize: function() {
 				this.$$('d2l-scroll-wrapper').notifyResize();
+			},
+			_applyClassNames: function() {
+
+				// don't allow a bunch of calls to queue up
+				if (this.__applyInQueue) return;
+				this.__applyInQueue = true;
+
+				fastdom.mutate(function() {
+					this.__applyInQueue = false;
+					var rows = this._getRowsAndCells();
+					for (var i = 0; i < rows.length; i++) {
+						var row = rows[i];
+						if (i === 0) {
+							row.elem.classList.add('d2l-table-row-first');
+						} else {
+							row.elem.classList.remove('d2l-table-row-first');
+						}
+						if (i === rows.length - 1) {
+							row.elem.classList.add('d2l-table-row-last');
+						} else {
+							row.elem.classList.remove('d2l-table-row-last');
+						}
+						for (var j = 0; j < row.cells.length; j++) {
+							var cell = row.cells[j];
+							if (j === 0) {
+								cell.classList.add('d2l-table-cell-first');
+							} else {
+								cell.classList.remove('d2l-table-cell-first');
+							}
+							if (j === row.cells.length - 1) {
+								cell.classList.add('d2l-table-cell-last');
+							} else {
+								cell.classList.remove('d2l-table-cell-last');
+							}
+						}
+					}
+				}.bind(this));
+
+			},
+			_handleLocalObserver: function() {
+				// descendants have been added/removed from somewhere in
+				// the <slot>,so re-calculate
+				this._applyClassNames();
+			},
+			_getRowsAndCells: function() {
```

Comment:
Maybe implement the same table API in the d2l-t* elements? Like `rows`, `cells` and write a `getTable` method that a common behavior uses so you can share code between d2l-table and d2l-table-wrapper?

Reply:
I considered it, but the two actually so different. The native `<table>` element exposes a `rows` collection and each row has a `cells` collection, so I can just walk those arrays. For `<d2l-table>` on the other hand, I need to recursively traverse the shadow DOM trees looking for `<d2l-row>` and `<d2l-cell>` elements, and I build up this collection as I go.